circuit/src/native.rs

```diff
@@ -0,0 +1,652 @@
+				let pk = filtered_set[i].0;
+				let ops_i = filtered_ops.get(&pk).unwrap_or_else(|| &default_op);
+
+				for j in 0..NUM_NEIGHBOURS {
+					let op = ops_i.scores[j].1 * s[i];
+					local_distr[j] = op;
+				}
+				distributions[i] = local_distr;
+			}
+
+			let mut new_s = [Fr::zero(); NUM_NEIGHBOURS];
+			for i in 0..NUM_NEIGHBOURS {
+				for j in 0..NUM_NEIGHBOURS {
+					new_s[i] += distributions[j][i];
+				}
+			}
+
+			s = new_s;
+		}
+
+		println!("new s: {:?}", s);
+
+		let mut sum = Fr::zero();
+		for x in s.iter() {
+			sum += x;
+		}
+		println!("sum: {:?}", sum);
+
+		s
+	}
+
+	fn filter_peers(
+		&self,
+	) -> (
+		[(PublicKey, Fr); NUM_NEIGHBOURS],
+		HashMap<PublicKey, Opinion>,
+	) {
+		let mut filtered_set: [(PublicKey, Fr); NUM_NEIGHBOURS] = self.set.clone();
+		let mut filtered_ops: HashMap<PublicKey, Opinion> = HashMap::new();
+
+		let mut valid_peers: Vec<PublicKey> = vec![];
+
+		// Convert the peer who didn't sign his opinion, to default peer
+		//
+		// Example:
+		//  In this set, peer3 didn't sign the opinion.
+		//		[(p1, 10), (p2, 10), (p3, 0)]
+		//   => [(p1, 10), (p2, 10), (null, 0)]
+		// Here, "null" means default peer
+		for i in 0..NUM_NEIGHBOURS {
+			let (pk, credits) = filtered_set[i].clone();
+			// If no credits, the peer did not sign his opinion
+			if credits == Fr::zero() {
+				filtered_set[i] = (PublicKey::default(), Fr::zero());
+			} else {
+				valid_peers.push(pk);
+			}
+		}
+
+		// Distribute the scores to valid peers
+		for i in 0..NUM_NEIGHBOURS {
+			let (pk_i, _) = filtered_set[i].clone();
+			if pk_i != PublicKey::default() {
+				let mut ops_i = self.ops.get(&pk_i).unwrap().clone();
+
+				// Give zero score for peers
+				//   - giving scores to itself
+				// 	 - giving scores to default peer
+				//
+				// Example:
+				//   Peer1 opinion
+				//		[(p1, 10), (p2, 10), (null, 10)]
+				// 	 => [(p1,  0), (p2, 10), (null,  0)]
+				//
+				for j in 0..NUM_NEIGHBOURS {
+					let (pk_j, score) = ops_i.scores[j].clone();
+					let true_pk = if i == j {
+						pk_i
+					} else if pk_j == pk_i {
+						PublicKey::default()
+					} else {
+						pk_j
+					};
+					let true_score = if true_pk == pk_i || true_pk == PublicKey::default() {
+						Fr::zero()
+					} else {
+						score
+					};
+					ops_i.scores[j] = (true_pk, true_score);
+				}
+
+				// Update the opinion array - pairs of (key, score)
+				//
+				// Example 1:
+				// 	filtered_set => [p1, null, p3]
+				//	Peer1 opinion
+				// 		[(p1, 0), (p6, 10),  (p3, 10)]
+				//   => [(p1, 0), (null, 0), (p3, 10)]
+				//
+				// Example 2:
+				// 	filtered_set => [p1, p2, p3]
+				//	Peer1 opinion
+				// 		[(p1, 0), (p3, 10), (null, 0)]
+				//   => [(p1, 0), (p2, 0),  (p3, 0)]
+				let op_score_sum =
+					ops_i.scores.iter().fold(Fr::zero(), |acc, &(_, score)| acc + score);
+				if op_score_sum == Fr::zero() {
+					panic!("The peer {i} didn't give any score to other peers!");
+				}
```

Comment:
We shouldn't panic here. Instead we should distribute the scores to everyone equally

Reply:
Hmm, the example case of this scenario is like:

`Set = [p1, p2, p3]`
Peer1 opinion(op1) =` [(p1, 10), (p2, 0), (p3, 0)]`
Then, the op1 becomes` [(p1, 0), (p2, 0), (p3, 0)] `after first filtering.
It means that the `p1` didn't give any score to `p2` and `p3`.

Do you think we should still give the scores even when `p1` originally does not give score?

Reply:
Yes, I think we should

Reply:
Ok, here is update commit (https://github.com/eigen-trust/protocol/pull/154/commits/9850d1b72e707d73162413c95512ad37feaa5714)